File: hierarchical_clustering/hc.R
# hierarchical clustering by 2blam

# set working directory
setwd("~/Documents/Git/ML/hierarchical_clustering")

#read the dataset
dataset = read.csv("Mall_Customers.csv")
X = dataset[4:5]

#dendrogram
dendrogram = hclust(dist(X, method="euclidean"), method="ward.D")
plot(dendrogram, 
     main=paste("Dendrogram"),
     xlab="Customers",
     ylab="Euclidean distances")

#optimal clusters = 5

# fit hc
hc = hclust(dist(X, method="euclidean"), method="ward.D")
y_hc = cutree(hc, k=5) #cut tree with 5 groups

library(cluster)
clusplot(X, 
         y_hc, 
         lines=0, 
         shade=TRUE, 
         color=TRUE, 
         labels=2, 
         plotchar=FALSE, 
         span=TRUE, 
         main=paste("Clusters of clients"), 
         xlab="Annual income",
         ylab="Spending score")

File: hierarchical_clustering/hc.py
# hierarchical clustering by 2blam

# import the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# import dataset
dataset = pd.read_csv("Mall_Customers.csv")
X = dataset.iloc[:, [3, 4]].values

# dendrogram
import scipy.cluster.hierarchy as sch
dendrogram = sch.dendrogram(sch.linkage(X, method="ward"))
plt.title("Dendrogram")
plt.xlabel("Customers")
plt.ylabel("Euclidean distances")
plt.show()

#optimal clusters: 5
from sklearn.cluster import AgglomerativeClustering
hc = AgglomerativeClustering(n_clusters=5, affinity="euclidean", linkage="ward")
y_hc = hc.fit_predict(X)

#visualize
plt.scatter(X[y_hc==0, 0], X[y_hc==0, 1], s=100, c="red", label="Careful")
plt.scatter(X[y_hc==1, 0], X[y_hc==1, 1], s=100, c="blue", label="Standard")
plt.scatter(X[y_hc==2, 0], X[y_hc==2, 1], s=100, c="green", label="Target")
plt.scatter(X[y_hc==3, 0], X[y_hc==3, 1], s=100, c="cyan", label="Careless")
plt.scatter(X[y_hc==4, 0], X[y_hc==4, 1], s=100, c="magenta", label="Sensible")
plt.title("Clusters of clients")
plt.xlabel("Annual income (k$)")
plt.ylabel("Spending Score (1-100)")
plt.legend()
plt.show()